### PR TITLE
NIFI-9034: Allow multiple urls in http Elasticsearch processors

### PR DESCRIPTION
**Pull request by eylonronen**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Review by MikeThomsen — COMMENTED** [review]
On the surface, I don't see any explicit blockers for merging, but I am skeptical that OkHttp is going to use the multiple URLs the way that the Elastic client does. The Elastic client, AFAIK, queries the Elastic configuration nodes and figures out how to do the query coordination properly. What are your thoughts on this subject with regard to the OkHttp implementation we have?

**Comment by eylonronen** [comment]
Hi, from my experience with ES when communicating with the cluster you have a few options:
1. Communicating with one of the data nodes - this option is the least recommended as it's likely to cause very high CPU and heap usage and when ingesting/querying large amount of data the said data node is bound to crash
2. Using a coordinator node - What is means is to still communicate with a single node, but one that does not store data, which means that it's load is smaller, yet here as well, with enough data (unfortunately i have been there) it to crashes
3. Using an external load balancer such as Nginx or HAproxy - It is a legit option which guarantees zero crashing, unless your cluster is not large enough to handle that much querying/ingestion and in which case no way will work for you :(
4. Load balancing in the client - by far the best option because of the same pros that were introduced in the previous one with no need of maintaining another server(s) and another technology

Basically at large scale the only valid options are 3 and 4, so to spare the need of a third-party load balancer i think that this feature is important. Also, you might look at the changes i made and observe that i implemented round robin load balancing (using Iterables.cycle()) which means the okHttp client still works with a single url every time, just a different one :)

**Comment by ChrisSamo632** [comment]
Unit Tests to show the splitting of the property value and iteration through the URLs would be good to show the functionality and ensure it doesn't break in future

**Comment by eylonronen** [comment]
Hi @ChrisSamo632 
I don't understand where and how to preform those unit tests... All the unit tests I see are testing processor functionalities whereas the ones you requested are checking class ones
It's just that I'm not sure if I should create a new test file for AbstractElasticsearchHttpProcessor and there create a mock class inheriting it and then test and assert using Junit or do something else...
I would love some guidance here, thank you very much :)

**Comment by ChrisSamo632** [comment]
On the question from @MikeThomsen and response from @eylonronen, I'd say the main thing that's missing in this implementation is any error handling within the simple round-robin implementation in this PR.

The Elasticsearch Client will (optionally) sniff the Elasticsearch cluster for all available hosts/URLs at startup/on connection - this implementation here relies upon the NiFi user configuring all the current Elasticsearch hosts and keeping them updated if the cluster membership changes.

Also, the Elasticsearch client will temporarily stop trying to connect to a host it knows is not responding - this PR will instead result in operations failing for any hosts that go offline (for example) and the NiFi user will need to cater for that, e.g. with a Retry loop on failure

That's not to say that this approach isn't a potentially useful option for Elasticsearch cluster users, it certainly avoids the need for them to introduce a load balancer in front of Elasticsearch if they don't already have one, and it will mean that if the single host they were able to connect with previously were to go offline, then others being configured in the processors means operations would work at least some of the time.

Generally I'd recommend people using the Elasticsearch Client-based processors where possible, but these OkHttp based processors have some functionality that the newer ES Client ones don't currently, so they're still relevant and this update feels like it's probably going to be useful.

**Comment by ChrisSamo632** [comment]
@eylonronen there are Checkstyle failures with your changes, please try building the `nifi-elasticsearch-bundle` with a `mvn -Pcontrib-check clean verify` command and check the output for the checkstyle failures

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/FetchElasticsearchHttp.java [hunk 1]
```diff
@@ -214,7 +214,7 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
             logger.debug("Fetching {}/{}/{} from Elasticsearch", new Object[]{index, docType, docId});
 
             // read the url property from the context
-            final String urlstr = StringUtils.trimToEmpty(context.getProperty(ES_URL).evaluateAttributeExpressions().getValue());
+            final String urlstr = StringUtils.trimToEmpty(getESUrl());
```
Comment (ChrisSamo632) [inline nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/FetchElasticsearchHttp.java:217]: The `getESUrl()` method calls `trim()` on the value being returned, so do we really need `StringUtils.trimToEmpty` or `.trim()` calls in all the individual processors (I'd imagine not)?

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/AbstractElasticsearchHttpProcessor.java [hunk 2]
```diff
@@ -282,6 +291,14 @@ protected JsonNode parseJsonResponse(InputStream in) throws IOException {
         return mapper.readTree(in);
     }
 
+    @OnScheduled
+    public void setup(ProcessContext context) {
+        esUrlsIteratorAtomicReference.set(Iterables.cycle(Arrays.asList(context.getProperty(ES_URL).evaluateAttributeExpressions().getValue().split(","))).iterator());
```
Comment (ChrisSamo632) [inline nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/AbstractElasticsearchHttpProcessor.java:296]: Better to `trim()` the values before setting them in the `Iterables.cycle` rather than every time they're retrieved in `getESUrl` - could create a `stream` from the `List` and apply a `map` function to `String::trim` each element before `collect`ing back into a `List` (or other similar approach)

**Comment by ChrisSamo632** [comment]
> Hi @ChrisSamo632
> I don't understand where and how to preform those unit tests... All the unit tests I see are testing processor functionalities whereas the ones you requested are checking class ones
> It's just that I'm not sure if I should create a new test file for AbstractElasticsearchHttpProcessor and there create a mock class inheriting it and then test and assert using Junit or do something else...
> I would love some guidance here, thank you very much :)

Probably got a few options, which you've already touched on in your question:
* Create an `AbstractElasticsearchHttpProcessorTest.java` (or `.groovy`) class that tests the new `getESUrl` method with different inputs (e.g. a single URL, no URLs, multiple URLs with/without whitespaces that need trimming) and create a `MockElasticsearchHttpProcessor.java` class to use for the tests (see [TestAbstractKinesisRecordProcessor](https://github.com/apache/nifi/blob/main/nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/test/java/org/apache/nifi/processors/aws/kinesis/stream/record/TestAbstractKinesisRecordProcessor.java#L83) for such an example)
* Create an `abstract` version of `AbstractElasticsearchHttpProcessorTest.java` with appropriate `@Test` methods and make this the super class of all the test classes for the processors that use the updated `AbstractElasticsearchHttpProcessor` class (e.g. `FetchElasticsearchHttpTest`), then we can also be sure that each of these processors is configured to use the new method too - this is the kind of approach I've taken in https://github.com/apache/nifi/pull/5193, for example
